### PR TITLE
[WIP] Add kubevirt provider to IPI

### PR DESCRIPTION
**Pull request by nirarg**
**WORK IN PROGRESS**

### CODE REVIEW

File: pkg/types/kubevirt/defaults/platform.go [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package defaults
```
Comment (bardielle) [inline pkg/types/kubevirt/defaults/platform.go:1]: remove that file from that commit

File: pkg/types/kubevirt/machinepool.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	Memory string `json:"memory,omitempty"`
+
+	// StorageSize is the size of VM's boot volume
+	// Format: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/resource/quantity.go
+	// +optional
+	StorageSize string `json:"storageSize,omitempty"`
+
+	SourcePvcName string `json:"SourcePvcName,omitempty"`
+
+	SecretName string `json:"SecretName.omitempty"`
```
Comment (bardielle) [inline pkg/types/kubevirt/machinepool.go]: Call it the same as in KubevirtMachinerProviderSpec - UnderKubeconfigSecretName

File: pkg/types/kubevirt/machinepool.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+// on kubevirt.
+type MachinePool struct {
+	// CPU is the mount of cpus used
+	// +optional
+	CPU uint32 `json:"cpu,omitempty"`
+
+	// Memory is the size of a VM's memory.
+	// Format: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/resource/quantity.go
+	// +optional
+	Memory string `json:"memory,omitempty"`
```
Comment (bardielle) [inline pkg/types/kubevirt/machinepool.go:13]: RequestedMemory

File: pkg/types/kubevirt/machinepool.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
+package kubevirt
+
+// MachinePool stores the configuration for a machine pool installed
+// on kubevirt.
+type MachinePool struct {
+	// CPU is the mount of cpus used
+	// +optional
+	CPU uint32 `json:"cpu,omitempty"`
```
Comment (bardielle) [inline pkg/types/kubevirt/machinepool.go:8]: RequestedCPU

File: pkg/types/kubevirt/machinepool.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+	// Memory is the size of a VM's memory.
+	// Format: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/resource/quantity.go
+	// +optional
+	Memory string `json:"memory,omitempty"`
+
+	// StorageSize is the size of VM's boot volume
+	// Format: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/resource/quantity.go
+	// +optional
+	StorageSize string `json:"storageSize,omitempty"`
```
Comment (bardielle) [inline pkg/types/kubevirt/machinepool.go:18]: RequestedStorage

File: pkg/types/kubevirt/machinepool.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+	SourcePvcName string `json:"SourcePvcName,omitempty"`
+
+	SecretName string `json:"SecretName.omitempty"`
+
+	// TODO kubevirt add more params
```
Comment (bardielle) [inline pkg/types/kubevirt/machinepool.go]: Add more params and remove that comment

File: pkg/types/kubevirt/metadata.go [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package kubevirt
+
+// Metadata contains kubevirt metadata (e.g. for uninstalling the cluster).
+type Metadata struct {
+	KubeConfig    string `json:"kubeconfig"`
```
Comment (bardielle) [inline pkg/types/kubevirt/metadata.go]: UnderKubeconfig

File: pkg/types/kubevirt/platform.go [hunk 8]
```diff
@@ -0,0 +1,25 @@
+package kubevirt
+
+// Platform stores all the global configuration that all
+// machinesets use.
+type Platform struct {
+	// The path to the kubeconfig file of the underkube cluster
+	KubeConfig string `json:"kubevirt_kubeconfig"`
```
Comment (bardielle) [inline pkg/types/kubevirt/platform.go]: UnderKubeconfig

File: pkg/types/kubevirt/platform.go [hunk 9]
```diff
@@ -0,0 +1,25 @@
+package kubevirt
+
+// Platform stores all the global configuration that all
```
Comment (bardielle) [inline pkg/types/kubevirt/platform.go:3]: Change the comment to be: holds all the platform information

File: pkg/types/kubevirt/platform.go [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+	// The namespace in the underkube cluster to install the overkube in it
+	// Namespace string `json:"kubevirt_namespace"`
+
+	// DefaultMachinePlatform is the default configuration used when
+	// installing on kubevirt for machine pools which do not define their
+	// own platform configuration.
+	// Default will set the image field to the latest RHCOS image.
+	// +optional
+	DefaultMachinePlatform *MachinePool `json:"defaultMachinePlatform,omitempty"`
```
Comment (bardielle) [inline pkg/types/kubevirt/platform.go]: Update the comment

File: pkg/types/kubevirt/validation/machinepool.go [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+		if p.OSDisk.SizeGB <= 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("sizeGB"), p.OSDisk.SizeGB, "disk size must be positive"))
+		}
+	}
+
+	return allErrs
+}
+
+// ValidVMType returns true if the vmType is supported.
+func ValidVMType(vmType ovirt.VMType) bool {
```
Comment (bardielle) [inline pkg/types/kubevirt/validation/machinepool.go]: Remove that function

File: pkg/types/kubevirt/validation/machinepool.go [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+	for _, v := range supportedVMTypes() {
+		if vmType == v {
+			return true
+		}
+	}
+	return false
+}
+
+// supportedVMTypes returns a slice of all supported VMTypes.
+func supportedVMTypes() []ovirt.VMType {
```
Comment (bardielle) [inline pkg/types/kubevirt/validation/machinepool.go]: Remove that function too

File: pkg/types/kubevirt/validation/machinepool_test.go [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	"github.com/openshift/installer/pkg/types/ovirt"
+)
+
+func TestValidateMachinePool(t *testing.T) {
```
Comment (bardielle) [inline pkg/types/kubevirt/validation/machinepool_test.go:12]: Add a todo comment to add more unit-tests

File: pkg/types/kubevirt/validation/platform.go [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+	"github.com/openshift/installer/pkg/types/ovirt"
+	"github.com/openshift/installer/pkg/validate"
+)
+
+// ValidatePlatform checks that the specified platform is valid.
+func ValidatePlatform(p *ovirt.Platform, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if err := validate.UUID(p.ClusterID); err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("ovirt_cluster_id"), p.ClusterID, err.Error()))
```
Comment (bardielle) [inline pkg/types/kubevirt/validation/platform.go]: change it to be kubevirt_cluster_id

File: pkg/types/kubevirt/validation/platform.go [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+	}
+	if err := validate.UUID(p.StorageDomainID); err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("ovirt_storage_domain_id"), p.StorageDomainID, err.Error()))
```
Comment (bardielle) [inline pkg/types/kubevirt/validation/platform.go]: change it to be kubevirt_cluster_id

File: pkg/types/kubevirt/validation/platform.go [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+import (
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	"github.com/openshift/installer/pkg/types/ovirt"
+	"github.com/openshift/installer/pkg/validate"
+)
+
+// ValidatePlatform checks that the specified platform is valid.
+func ValidatePlatform(p *ovirt.Platform, fldPath *field.Path) field.ErrorList {
```
Comment (bardielle) [inline pkg/types/kubevirt/validation/platform.go]: update that function to be relevant to kubevirt platform

File: pkg/types/kubevirt/validation/platform_test.go [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	"github.com/openshift/installer/pkg/types/ovirt"
+)
+
+func validPlatform() *ovirt.Platform {
+	return &ovirt.Platform{
```
Comment (bardielle) [inline pkg/types/kubevirt/validation/platform_test.go]: update it with validations

File: pkg/types/kubevirt/validation/platform_test.go [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+		StorageDomainID:        "57e42205-02ac-46e1-a6d1-07459d94bc51",
+		VNICProfileID:          "57e42205-02ac-46e1-a6d1-07459d94bc52",
+		NetworkName:            "ocp-blue",
+		APIVIP:                 "10.0.0.1",
+		IngressVIP:             "10.0.0.3",
+		DefaultMachinePlatform: nil,
+	}
+}
+
+func TestValidatePlatform(t *testing.T) {
```
Comment (bardielle) [inline pkg/types/kubevirt/validation/platform_test.go:23]: Add a todo comment to add more tests

File: pkg/types/ovirt/doc.go [hunk 19]
```diff
@@ -3,4 +3,4 @@
 package ovirt
 
 // Name is the name for the ovirt platform.
-const Name string = "ovirt"
+const 	Name string = "ovirt"
```
Comment (bardielle) [inline pkg/types/ovirt/doc.go]: Why did you change that file?

File: pkg/tfvars/kubevirt/kubevirt.go [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+	"encoding/json"
+
+	v1 "github.com/kubevirt/cluster-api-provider-kubevirt/pkg/apis/kubevirtprovider/v1alpha1"
+	// "github.com/openshift/installer/pkg/rhcos"
+	// "github.com/openshift/installer/pkg/tfvars/internal/cache"
+)
+
+type config struct {
+	Namespace        string `json:"kubevirt_namespace"`
+	Kubeconfig       string `json:"kubevirt_kubeconfig"`
```
Comment (bardielle) [inline pkg/tfvars/kubevirt/kubevirt.go]: UnderKubeconfig

File: pkg/tfvars/kubevirt/kubevirt.go [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+	ImageURL         string `json:"kubevirt_image_url"`
+	SourcePvcName    string `json:"kubevirt_source_pvc_name"`
+	SecretName       string `json:"kubevirt_secret_name"`
+	RequestedMemory  string `json:"kubevirt_master_memory"`
+	RequestedCPU     uint32 `json:"kubevirt_master_cpu"`
+	RequestedStorage string `json:"kubevirt_master_storage"`
+	MachineType      string `json:"kubevirt_master_machine_type"`
```
Comment (bardielle) [inline pkg/tfvars/kubevirt/kubevirt.go]: Remove that type

File: pkg/tfvars/kubevirt/kubevirt.go [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+	// "github.com/openshift/installer/pkg/rhcos"
+	// "github.com/openshift/installer/pkg/tfvars/internal/cache"
+)
+
+type config struct {
+	Namespace        string `json:"kubevirt_namespace"`
+	Kubeconfig       string `json:"kubevirt_kubeconfig"`
+	ImageURL         string `json:"kubevirt_image_url"`
+	SourcePvcName    string `json:"kubevirt_source_pvc_name"`
+	SecretName       string `json:"kubevirt_secret_name"`
```
Comment (bardielle) [inline pkg/tfvars/kubevirt/kubevirt.go]: SecretName of what? I think you need to remove it

File: pkg/tfvars/kubevirt/kubevirt.go [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+	RequestedCPU     uint32 `json:"kubevirt_master_cpu"`
+	RequestedStorage string `json:"kubevirt_master_storage"`
+	MachineType      string `json:"kubevirt_master_machine_type"`
+	StorageClass     string `json:"kubevirt_storage_class"`
+}
+
+// TFVarsSources contains the parameters to be converted into Terraform variables
+type TFVarsSources struct {
+	MasterSpecs []*v1.KubevirtMachineProviderSpec
+	Kubeconfig  string
```
Comment (bardielle) [inline pkg/tfvars/kubevirt/kubevirt.go]: UnderKubeconfig

File: pkg/tfvars/kubevirt/kubevirt.go [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+	ImageURL    string
```
Comment (bardielle) [inline pkg/tfvars/kubevirt/kubevirt.go]: ImageUrl of what? rchoseimageUrl

File: pkg/terraform/exec/plugins/kubernetes-alpha.go [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+	"context"
+
+	goplugin "github.com/hashicorp/go-plugin"
+	"github.com/hashicorp/terraform-plugin-sdk/plugin"
+	"github.com/hashicorp/terraform-plugin-sdk/tfplugin5"
+	"github.com/hashicorp/terraform-provider-kubernetes-alpha/provider"
+	"google.golang.org/grpc"
+)
+
+// func init() {
```
Comment (bardielle) [inline pkg/terraform/exec/plugins/kubernetes-alpha.go]: remove that comment

File: pkg/terraform/exec/plugins/kubernetes-alpha.go [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+// 	provider.Serve()
+// }
+
+func init() {
+	defer provider.InitDevLog()()
+	handshake := goplugin.HandshakeConfig{
+		ProtocolVersion: 5,
+		// The magic cookie values should NEVER be changed.
+		MagicCookieKey:   "TF_PLUGIN_MAGIC_COOKIE",
+		MagicCookieValue: "d602bf8f470bc67ca7faa0386276bbdd4330efaf76d1a219cb4d6991ca9872b2",
```
Comment (bardielle) [inline pkg/terraform/exec/plugins/kubernetes-alpha.go]: Use consts

File: pkg/terraform/exec/plugins/kubernetes-alpha.go [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+	"google.golang.org/grpc"
+)
+
+// func init() {
+// 	defer provider.InitDevLog()()
+// 	provider.Dlog.Printf("Starting up")
+// 	provider.Serve()
+// }
+
+func init() {
```
Comment (bardielle) [inline pkg/terraform/exec/plugins/kubernetes-alpha.go]: Add a comment of description

File: pkg/terraform/exec/plugins/kubernetes.go [hunk 28]
```diff
@@ -0,0 +1,14 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/plugin"
+	"github.com/hashicorp/terraform-provider-kubernetes/kubernetes"
+)
+
+func init() {
+	exec := func() {
+		plugin.Serve(&plugin.ServeOpts{
+			ProviderFunc: kubernetes.Provider})
+	}
+	KnownPlugins["terraform-provider-kubernetes"] = exec
```
Comment (bardielle) [inline pkg/terraform/exec/plugins/kubernetes.go:13]: use consts

File: pkg/asset/installconfig/basedomain.go [hunk 29]
```diff
@@ -61,6 +62,9 @@ func (a *baseDomain) Generate(parents asset.Parents) error {
 		if !(gcpconfig.IsForbidden(err) || gcpconfig.IsThrottled(err)) {
 			return err
 		}
+	case kubevirt.Name:
+		a.BaseDomain = "svc.cluster.local"
```
Comment (bardielle) [inline pkg/asset/installconfig/basedomain.go]: Use consts

File: pkg/asset/installconfig/basedomain.go [hunk 30]
```diff
@@ -61,6 +62,9 @@ func (a *baseDomain) Generate(parents asset.Parents) error {
 		if !(gcpconfig.IsForbidden(err) || gcpconfig.IsThrottled(err)) {
 			return err
 		}
+	case kubevirt.Name:
```
Comment (bardielle) [inline pkg/asset/installconfig/basedomain.go]: Add a todo -use a variable from user

File: pkg/asset/installconfig/installconfig.go [hunk 31]
```diff
@@ -187,5 +199,8 @@ func (a *InstallConfig) platformValidation() error {
 	if a.Config.Platform.OpenStack != nil {
 		return icopenstack.Validate(a.Config)
 	}
+	if a.Config.Platform.Kubevirt != nil {
```
Comment (bardielle) [inline pkg/asset/installconfig/installconfig.go:200]: change it to switch-case?

File: pkg/asset/installconfig/kubevirt/config.go [hunk 32]
```diff
@@ -0,0 +1,11 @@
+package kubevirt
+
+import "io/ioutil"
+
+func LoadUnderKubeConfig(filePath string) ([]byte, error) {
```
Comment (bardielle) [inline pkg/asset/installconfig/kubevirt/config.go]: LoadUnderKubeconfig

File: pkg/terraform/exec/plugins/kubernetes.go (hunk 28 above)
Comment (bardielle) [reply pkg/terraform/exec/plugins/kubernetes.go:13]: still relevant

File: pkg/terraform/exec/plugins/kubernetes-alpha.go (hunk 26 above)
Comment (bardielle) [reply pkg/terraform/exec/plugins/kubernetes-alpha.go]: still relevant

File: data/data/kubevirt/main.tf [hunk 33]
```diff
@@ -0,0 +1,68 @@
+provider "kubernetes" {
+  config_path = "/home/nargaman/.kube/config"
```
Comment (bardielle) [inline data/data/kubevirt/main.tf]: change the path to be relative for the running directory

File: data/data/kubevirt/main.tf [hunk 34]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+provider "kubernetes-alpha" {
+  server_side_planning = false
+  config_path = "/home/nargaman/.kube/config"
```
Comment (bardielle) [inline data/data/kubevirt/main.tf]: change the path to be relative for the running directory

File: pkg/asset/installconfig/kubevirt/kubevirt.go [hunk 35]
```diff
@@ -0,0 +1,160 @@
...
+		IngressVIP:               ingressVIP,
+		NetworkName:              networkName,
+	}, nil
+}
+
+func selectKubeConfig() (string, error) {
+	var selectedKubeConfig string
+
+	defaultValue := ""
+	home := os.Getenv("HOME")
```
Comment (bardielle) [inline pkg/asset/installconfig/kubevirt/kubevirt.go]: In my opiniion you need to take the kubeconfig from a relative place in the running directory

File: data/data/manifests/openshift/role-cloud-creds-secret-reader.yaml.template [hunk 36]
```diff
@@ -14,6 +14,8 @@ metadata:
   name: vsphere-creds-secret-reader
 {{- else if .CloudCreds.Ovirt}}
   name: ovirt-creds-secret-reader
+{{- else if .CloudCreds.Kubevirt}}
+  name: kubevirt-creds-secret-reader
```
Comment (bardielle) [inline data/data/manifests/openshift/role-cloud-creds-secret-reader.yaml.template:18]: what is it for?

File: pkg/asset/ignition/machine/node.go [hunk 37]
```diff
@@ -36,6 +37,8 @@ func pointerIgnitionConfig(installConfig *types.InstallConfig, rootCA []byte, ro
 		if installConfig.VSphere.APIVIP != "" {
 			ignitionHost = net.JoinHostPort(installConfig.VSphere.APIVIP, "22623")
 		}
+	case kubevirttypes.Name:
+		ignitionHost = net.JoinHostPort(installConfig.Kubevirt.APIVIP, "22623")
```
Comment (bardielle) [inline pkg/asset/ignition/machine/node.go:41]: I think that this section could be refactored: 
`switch installConfig.Platform.Name() {
case baremetaltypes.Name:
	apiVipParam = installConfig.BareMetal.APIVIP
....
}
ignitionHost = net.JoinHostPort( apiVipParam , "22623")
`


But I'm not sure

File: data/data/kubevirt/main.tf (hunk 33 above)
Comment (nirarg) [reply data/data/kubevirt/main.tf]: Done

File: data/data/kubevirt/main.tf (hunk 34 above)
Comment (nirarg) [reply data/data/kubevirt/main.tf]: Done

File: pkg/asset/installconfig/kubevirt/kubevirt.go (hunk 35 above)
Comment (nirarg) [reply pkg/asset/installconfig/kubevirt/kubevirt.go]: Done
